include/nbl/builtin/hlsl/macros.hlsl

```diff
@@ -0,0 +1,37 @@
+// Copyright (C) 2022 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_MACROS_INCLUDED_
+#define _NBL_BUILTIN_HLSL_MACROS_INCLUDED_
+
+#define NBL_HLSL_EVAL(X) X
+
+#define NBL_HLSL_EQUAL(X,Y) (NBL_HLSL_EVAL(X)==NBL_HLSL_EVAL(Y))
+#define NBL_HLSL_NOT_EQUAL(X,Y) (NBL_HLSL_EVAL(X)!=NBL_HLSL_EVAL(Y))
+
+#define NBL_HLSL_LESS(X,Y) (NBL_HLSL_EVAL(X)<NBL_HLSL_EVAL(Y))
+#define NBL_HLSL_GREATER(X,Y) (NBL_HLSL_EVAL(X)>NBL_HLSL_EVAL(Y))
+
+#define NBL_HLSL_IS_NOT_POT(v) (NBL_HLSL_EVAL(v)&(NBL_HLSL_EVAL(v)-1))!=0
+#define NBL_HLSL_IS_POT(v) (!NBL_HLSL_IS_NOT_POT(v))
+
+#define NBL_HLSL_CONCAT_IMPL2(X,Y) X ## Y
+#define NBL_HLSL_CONCAT_IMPL(X,Y) NBL_HLSL_CONCAT_IMPL2(X,Y)
+#define NBL_HLSL_CONCATENATE2(X,Y) NBL_HLSL_CONCAT_IMPL(NBL_HLSL_EVAL(X),NBL_HLSL_EVAL(Y))
+#define NBL_HLSL_CONCATENATE3(X,Y,Z) NBL_HLSL_CONCATENATE2(NBL_HLSL_CONCATENATE2(X,Y),NBL_HLSL_EVAL(Z))
+#define NBL_HLSL_CONCATENATE4(X,Y,Z,W) NBL_HLSL_CONCATENATE2(NBL_HLSL_CONCATENATE2(X,Y),NBL_HLSL_CONCATENATE2(Z,W))
+
+#define NBL_HLSL_AND(X,Y) (NBL_HLSL_EVAL(X)&NBL_HLSL_EVAL(Y))
+#define NBL_HLSL_OR(X,Y) (NBL_HLSL_EVAL(X)|NBL_HLSL_EVAL(Y))
+#define NBL_HLSL_XOR(X,Y) (NBL_HLSL_EVAL(X)^NBL_HLSL_EVAL(Y))
+
+#define NBL_HLSL_ADD(X,Y) (NBL_HLSL_EVAL(X)+NBL_HLSL_EVAL(Y))
+#define NBL_HLSL_SUB(X,Y) (NBL_HLSL_EVAL(X)-NBL_HLSL_EVAL(Y))
+#define NBL_HLSL_MUL(X,Y) (NBL_HLSL_EVAL(X)*NBL_HLSL_EVAL(Y))
+
+// https://github.com/google/shaderc/issues/1155
+// This issue still seems to exist
+// #define NBL_HLSL_MAX(X,Y) (((NBL_HLSL_EVAL(X))>(NBL_HLSL_EVAL(Y))) ? (NBL_HLSL_EVAL(X)):(NBL_HLSL_EVAL(Y)))
+// #define NBL_HLSL_MIN(X,Y) (((NBL_HLSL_EVAL(X))<(NBL_HLSL_EVAL(Y))) ? (NBL_HLSL_EVAL(X)):(NBL_HLSL_EVAL(Y)))
```

Comment:
can you check if it works if you slap a big `NBL_HLSL_EVAL` around the `< ? :`

Reply:
Tried this in compiler explorer

`int a = NBL_HLSL_EVAL(NBL_HLSL_MAX(5,4));`

and it preprocesses to

`int a = (((5)>(4)) ? (5):(4));`  👎

Reply:
can you try and see if DXC will complain about `uint someArray[NBL_HLSL_MAX(5,4)]` ?

Reply:
Doesn't complain and it actually seems to work. The DXIL IR contains the following
```
!9 = !DIGlobalVariable(name: "someArray", linkageName: "\01?someArray@@3QBIB", scope: !0, file: !1, line: 39, type: !10, isLocal: false, isDefinition: true)
!10 = !DICompositeType(tag: DW_TAG_array_type, baseType: !11, **size: 160**, align: 32, elements: !14)
```
The `size:160` matches the 5 uints the array is supposed to be.

Reply:
but 5 uints is 20 bytes ?

Reply:
ah okay, its in bits :D

Reply:
we can uncomment NBL_HLSL_MIN and MAX then